proposals/1533-host-specific-tasklist.md

```diff
@@ -0,0 +1,241 @@
+
+* For RecreateSession(), the same thing happens. The only difference is that the session creation activity will be **scheduled on the resource specific task list instead of a global one**.
+
+* For CompleteSession(), another special **completion activity** will be scheduled on the process specific tasklist. The sole purpose of this activity is to stop the corresponding creation activity created by the CreateSession()/CreateSessionWithHostID() API call, so that the resource used by the session can be released. Note that, **the completion activity must be scheduled on the resource specific tasklist (not a global one)** since we need to make sure the completion activity and the creation activity it needs to stop are running by the same worker.
+
+### Activity Worker
+
+When EnableSessionWorker is set to true, two more activity workers will be started.
+
+* Session Creation Worker: this worker polls from a global tasklist whose name is derived from the tasklist name specified in the workerOption. This worker executes only one kind of activity: the special session creation activity. This special activity serves several purposes:
+
+   1. **Send the resource specific tasklist name** back to the workflow through signal.
+
+   2. Keep **heart beating** to cadence server till the end of the session.
+
+      If the heart beating returns EntityNotExist Error, this means the workflow/session has gone and this special activity should stop to release resource.
+
+      If on the other hand, the worker is down, cadence server will send heartbeat timeout error to the workflow, so the workflow can be notified and update the session state. Later ExecuteActivity() call with the failed session context will also fail.
+
+   3. Check the number of currently running sessions on the worker. If the maximum number of concurrent session has been reached, return an error to fail the creation.
+
+   One detail we need to ensure is that when the maximum number of concurrent session has been reached, this worker should stop polling from the global task list.
+
+* Session Activity Worker: this worker only polls from the resource specific tasklist. There are three kinds of activities can be run by this worker.
+
+   1. User activities in a session that belongs to this host.
+
+   2. Special activity to complete a session.
+
+      The sole purpose of this activity is to stop the long running session creation activity so that new sessions can be executed.
+
+   3. Special session creation activity scheduled by RecreateSession().
+
+When domain failover happens, since no worker will poll from the resource specific task, all activities within a session will timeout and user will get an error. Also cadence server will detect heartbeat timeout from the session creation activity.
```

Comment:
and user will get -> and the user will get
cadence server -> Cadence server
detect heartbeat -> detect the heartbeat